lib/rules/no-ancestor-directory-import.js

```diff
@@ -0,0 +1,49 @@
+const {extname, basename, relative, sep} = require('path');
+const resolve = require('eslint-module-utils/resolve').default;
+
+module.exports = {
+  meta: {
+    docs: {
+      description:
+        'Prefer that imports from within a directory extend to the file from where they are importing without relying on an index file.',
+      category: 'Best Practices',
+      recommended: false,
+      uri:
+        'https://github.com/Shopify/eslint-plugin-shopify/blob/master/docs/rules/no-ancestor-directory-import.md',
+    },
+    fixable: null,
+  },
+  create(context) {
+    function isAncestorDirectoryImport(resolvedSource) {
+      const relativeDifference = relative(
+        context.getFilename(),
+        resolvedSource,
+      );
+
+      const parts = relativeDifference
+        .split(sep)
+        .filter((part) => part[0] !== '.');
+
+      if (parts.length > 1) {
+        return false;
+      }
+
+      return basename(parts[0], extname(parts[0])) === 'index';
+    }
+
+    return {
+      ImportDeclaration(node) {
+        const importSource = node.source.value;
+        const resolvedSource = resolve(importSource, context);
+
+        if (resolvedSource && isAncestorDirectoryImport(resolvedSource)) {
+          context.report({
+            node,
+            message:
+              'Relative local imports should extend to the file from where they are importing without relying on an index file.',
```

Comment:
I don't know if this message is quite right; since we resolve the source, it might not even be a relative import (i.e., something in `app/sections` asking for `index` would fail this rule, as that likely resolves to `app/index.js`